ppx/ppx_tyxml.ml

```diff
@@ -49,29 +49,105 @@ module Loc = struct
+      the given strings, using [Markup.preprocess_input_stream]. This is when
+      Markup.ml assigns locations to each character.
+
+      [make_location_map] then drains the parser input stream. As the input
+      string list is consumed, whenever a new string is taken, its OCaml
+      location and the current Markup.ml location are noted in a list
+      [location_map].
+
+      [make_location_map] then creates a function that behaves as follows: when
+      given a Markup.ml location [l], it performs a search in [location_map] for
+      the greatest Markup.ml location [l'] preceding [l]. Then, it uses the
+      OCaml location corresponding to [l'] to compute an OCaml location
+      corresponding to [l].
+
+      The search is a linear search, because the list of locations is not
+      expected to be large, and the query happens at most once per PPX run, on
+      occurrence of an error. *)
+  let make_location_map located_strings =
+    (* [source] is a byte stream created from the string list, which calls
+       [!starting_a_string] each time it moves on to a new string in the
+       list. *)
+    let starting_a_string = ref (fun _ -> ()) in
+    let source =
+      let strings = ref located_strings in
+      let offset = ref 0 in
+
+      let rec next_byte () = match !strings with
+        | [] -> None
+        | (s, loc)::rest ->
+          if !offset = 0 then !starting_a_string loc;
+
+          if !offset < String.length s then begin
+            offset := !offset + 1;
+            Some (s.[!offset - 1])
+          end
+          else begin
+            offset := 0;
+            strings := rest;
+            next_byte ()
+          end
+      in
+
+      Markup.fn next_byte
     in
-    let line = loc.pos_lnum + line - 1 in
 
-    let position =
-      {pos_fname = loc.pos_fname;
-       pos_lnum  = line;
-       pos_bol   = 0;
-       pos_cnum  = column};
+    (* Use Markup.ml to assign locations to characters in [source], and note
+       the Markup.ml and OCaml locations of the start of each string. *)
+    let preprocessed_input_stream, get_markupml_location =
+      source
+      |> Markup.Encoding.decode Markup.Encoding.utf_8
+      |> Markup.preprocess_input_stream
     in
 
-    one_width position
+    let location_map = ref [] in
+    starting_a_string := begin fun ocaml_position ->
+      location_map := (get_markupml_location (), ocaml_position)::!location_map
+    end;
+
+    Markup.drain preprocessed_input_stream;
+    let location_map = List.rev !location_map in
+
+    (* The function proper which translates Markup.ml locations into OCaml
+       locations. *)
+    fun given_markup_location ->
+      let rec bounded_maximum best = function
+        | [] -> best
+        | ((noted_markup_location, _) as loc)::rest ->
+          if Markup.compare_locations
+               noted_markup_location given_markup_location > 0 then best
+          else bounded_maximum (Some loc) rest
```

Comment:
If I understand correctly, `best` here is the last markup location that was smaller ? Could you change the name to reflect that ?